include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,222 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+#pragma once
+#ifndef __WIL_CPPWINRT_AUTHORING_INCLUDED
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED
+
+namespace wil
+{
+    template <typename T>
+    struct read_only_property
+    {
+        read_only_property(T value) : m_value(value) { }
+
+        operator T() const noexcept { return m_value; }
+        const T& operator()() const noexcept { return m_value; }
+
+        read_only_property(const read_only_property& other) noexcept : m_value(other.m_value) { }
+        read_only_property(read_only_property&& other) noexcept : m_value(std::move(other.m_value)) { }
+
+        template<typename... TArgs>
+        read_only_property(TArgs&&... args) : m_value(std::forward<TArgs>(args)...) { }
+
+    protected:
+        T m_value;
+    };
+
+    template <typename T>
+    struct read_write_property : read_only_property<T>
+    {
+        read_write_property(T value) : read_only_property<T>(value) { }
+
+        template<typename... TArgs>
+        read_write_property(TArgs&&... args) : read_only_property<T>(std::forward<TArgs>(args)...) { }
+
+        using read_only_property<T>::operator T;
+        using read_only_property<T>::operator();
+        read_write_property& operator()(const T& value) noexcept
+        {
+            this->m_value = value;
+            return *this;
+        }
+
+        read_write_property& operator=(T value) noexcept
+        {
+            this->m_value = value;
+            return *this;
+        }
+
+        template<typename U, typename = std::enable_if_t<std::is_convertible_v<U, T>>>
+        read_write_property& operator=(const read_only_property<U>& other) noexcept
+        {
+            this->m_value = other.m_value;
+            return *this;
+        }
+
+        template<typename U, typename = std::enable_if_t<std::is_convertible_v<U, T>>>
+        read_write_property& operator=(read_write_property<U>&& other) noexcept
+        {
+            this->m_value = std::move(other.m_value);
+            return *this;
+        }
+    };
+
+#pragma region WinRT / XAML helpers
+#ifdef WINRT_Windows_Foundation_H

```

Comment:
If we're going to persist with this pattern, then we need to at least ensure that the header can be included multiple times so that functionality can "light up"